x-pack/plugins/alerts/server/authorization/alerts_authorization.ts

```diff
@@ -0,0 +1,262 @@
+  auditLogger: AlertsAuthorizationAuditLogger;
+  authorization?: SecurityPluginSetup['authz'];
+}
+
+export class AlertsAuthorization {
+  private readonly alertTypeRegistry: AlertTypeRegistry;
+  private readonly features: FeaturesPluginStart;
+  private readonly request: KibanaRequest;
+  private readonly authorization?: SecurityPluginSetup['authz'];
+  private readonly auditLogger: AlertsAuthorizationAuditLogger;
+
+  constructor({
+    alertTypeRegistry,
+    request,
+    authorization,
+    features,
+    auditLogger,
+  }: ConstructorOptions) {
+    this.request = request;
+    this.authorization = authorization;
+    this.features = features;
+    this.alertTypeRegistry = alertTypeRegistry;
+    this.auditLogger = auditLogger;
+  }
+
+  public async ensureAuthorized(alertTypeId: string, consumer: string, operation: string) {
+    const { authorization } = this;
+    if (authorization) {
+      const alertType = this.alertTypeRegistry.get(alertTypeId);
+      const requiredPrivilegesByScope = {
+        consumer: authorization.actions.alerting.get(alertTypeId, consumer, operation),
+        producer: authorization.actions.alerting.get(alertTypeId, alertType.producer, operation),
+      };
+
+      // We special case the Alerts Management `consumer` as we don't want to have to
+      // manually authorize each alert type in the management UI
+      const shouldAuthorizeConsumer = consumer !== AlertsFeatureId;
+
+      const checkPrivileges = authorization.checkPrivilegesDynamicallyWithRequest(this.request);
+      const { hasAllRequested, username, privileges } = await checkPrivileges(
+        shouldAuthorizeConsumer && consumer !== alertType.producer
+          ? [
+              // check for access at consumer level
+              requiredPrivilegesByScope.consumer,
+              // check for access at producer level
+              requiredPrivilegesByScope.producer,
+            ]
+          : [
+              // skip consumer privilege checks under `alerts` as all alert types can
+              // be created under `alerts` if you have producer level privileges
+              requiredPrivilegesByScope.producer,
+            ]
+      );
+
+      if (hasAllRequested) {
+        this.auditLogger.alertsAuthorizationSuccess(
+          username,
+          alertTypeId,
+          ScopeType.Consumer,
+          consumer,
+          operation
+        );
+      } else {
+        const authorizedPrivileges = pluck(
+          privileges.filter((privilege) => privilege.authorized),
+          'privilege'
+        );
+        const unauthorizedScopes = mapValues(
+          requiredPrivilegesByScope,
+          (privilege) => !authorizedPrivileges.includes(privilege)
+        );
+
+        const [unauthorizedScopeType, unauthorizedScope] =
+          shouldAuthorizeConsumer && unauthorizedScopes.consumer
+            ? [ScopeType.Consumer, consumer]
+            : [ScopeType.Producer, alertType.producer];
+
+        throw Boom.forbidden(
```

Comment:
Haha well, I'm aligning it with the existing code so that we remove the mall at once - otherwise you'll get a weird mix of some code wit hBoom and some without.

That why we have this issue: https://github.com/elastic/kibana/issues/49822